components/edit.js

```diff
@@ -0,0 +1,235 @@
+          </li>
+        ))}
+      </ul>
+    </div>
+  )
+}
+
+SavedErrors.propTypes = {
+  errors: PropTypes.array.isRequired,
+}
+
+const EditPage = ({ item }) => {
+  const path = item["@id"]
+
+  const loggedIn = useAuthenticated()
+
+  const editable = (session, item) => {
+    // cannot edit if not logged in or object not editable
+    const editable = canEdit(item)
+    return loggedIn && editable
+  }
+
+  const jsonErrors = (json) => {
+    // cannot save if we cannot edit or if the JSON is wrong
+    try {
+      JSON.parse(json)
+      return []
+    } catch (err) {
+      // Save the error
+      return [err.message]
+    }
+  }
+
+  const { session } = useContext(SessionContext)
+
+  const [text, setText] = useState(() => {
+    JSON.stringify({}, null, 4)
+  })
+
+  const [saveErrors, setErrors] = useState([])
+
+  const [editorStatus, setEditorStatus] = useState({
+    canEdit: editable(session, item),
+    canSave: editable(session, item),
+    errors: [],
+  })
+
+  useEffect(() => {
+    const fetch = new Fetch(session)
+    fetch.getObject(`${path}?frame=edit`, "GET").then((value) => {
+      const json = value.json()
+      return json
+    }).then((value) => {
+      setText(JSON.stringify(sortedJson(value), null, 4))
+    })
+  }, [path, session])
+
+  const router = useRouter()
+
+  const onChange = (newValue) => {
+    setText(newValue)
+    const errors = jsonErrors(newValue)
+    const status = {
+      canEdit: editable(session, item),
+      canSave: errors.length == 0 && editable(session, item),
+      errors,
+    }
+    setEditorStatus(status)
+  }
+
+  const save = () => {
+    setEditorStatus({
+      canEdit: false,
+      canSave: false,
+      errors: [],
+    })
+    const value = sortedJson(JSON.parse(text))
+    const fetch = new Fetch(session)
+    fetch.updateObject(path, "PATCH", value).
+    then((response) => {
+      if (response.ok) {
+        setErrors([])
+        router.push(path)
+      } else {
+        setEditorStatus({
+          canEdit: true,
+          canSave: true,
+          errors: [],
+        })
+        const errors = response.errors.map((err) => ({
+          description: err.description,
+          keys: err.names
+            .map((val) => {
+              `\`${val}\``
+            })
+            .join(", "),
+        }))
+        setErrors(errors)
+      }
+    })
+  }
+
+  return (
+    <div className="space-y-1">
+      <EditJson
+        text={text}
+        onChange={onChange}
+        enabled={editorStatus.canEdit}
+        errors={editorStatus.errors}
+      />
+      <div>
+        <SaveCancelControl
+          cancelClick={() => ({})}
+          saveClick={save}
+          itemPath={path}
+          saveEnabled={editorStatus.canSave}
+        />
+        {saveErrors.length > 0 && <SavedErrors errors={saveErrors} />}
+      </div>
+    </div>
+  )
+}
+
+EditPage.propTypes = {
+  item: PropTypes.object.isRequired,
+}
+
+export default EditPage
```

Comment:
This file isn’t prettier compliant. The .then() formatting changes when you save the file.